lib/Dialect/FIRRTL/Transforms/CheckWidths.cpp

```diff
@@ -0,0 +1,106 @@
+//===- CheckWidths.cpp - Check that width of types are known ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the CheckWidths pass.
+//
+// Modeled after:
+// https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/passes/CheckWidths.scala
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+/// A simple pass that emits errors for any occurrences of `uint`, `sint`, or
+/// `analog` without a known width.
+class CheckWidthsPass : public CheckWidthsBase<CheckWidthsPass> {
+  void runOnOperation() override;
+  void checkType(Type type, Type reportType, Location loc,
+                 llvm::function_ref<void(InFlightDiagnostic &)> addContext);
+
+  InFlightDiagnostic emitError(Location loc, const Twine &message) {
+    signalPassFailure();
+    return mlir::emitError(loc, message);
+  }
+};
+} // namespace
+
+void CheckWidthsPass::runOnOperation() {
+  getOperation()->walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
```

Comment:
By running on firrtl.module ops you can pull the port checking out of the walk pass.